core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -284,12 +289,18 @@ class KafkaController(val config : KafkaConfig, zkUtils: ZkUtils, val brokerStat
                   }
                   // If the broker is a follower, updates the isr in ZK and notifies the current leader
                   replicaStateMachine.handleStateChanges(Set(PartitionAndReplica(topicAndPartition.topic,
-                    topicAndPartition.partition, id)), OfflineReplica)
+                    topicAndPartition.partition, id)), OfflineReplica, newBatch = false)
                 }
               }
             }
           }
       }
+
+      inLock(controllerContext.controllerLock) {
```

Comment:
Hmm, could you explain a bit more why the improvement could be that significant (2-3 mins to 15 seconds)? Intuitively, each LeaderAndIsrRequest and UpdateMetadataRequest requires 1 roundtrip of network communication. Each write to ZK requires at least 2 roundtrip of network communication (1 to the ZK leader and another from the leader to the follower).  So, for each partition, the bulk of the cost would probably be 1 + 1 + 2 = 4 roundtrip of network communication. This patch eliminates the first 2. So, the saving should be no more than 50%. Was the broker's request handler pool overloaded at that time?